astronomy/date_body.hpp

```diff
@@ -0,0 +1,781 @@
+    +0, +1,  // 1978
+    +0, +1,  // 1979
+    +0, +0,  // 1980
+    +1, +0,  // 1981
+    +1, +0,  // 1982
+    +1, +0,  // 1983
+    +0, +0,  // 1984
+    +1, +0,  // 1985
+    +0, +0,  // 1986
+    +0, +1,  // 1987
+    +0, +0,  // 1988
+    +0, +1,  // 1989
+    +0, +1,  // 1990
+    +0, +0,  // 1991
+    +1, +0,  // 1992
+    +1, +0,  // 1993
+    +1, +0,  // 1994
+    +0, +1,  // 1995
+    +0, +0,  // 1996
+    +1, +0,  // 1997
+    +0, +1,  // 1998
+    +0, +0,  // 1999
+    +0, +0,  // 2000
+    +0, +0,  // 2001
+    +0, +0,  // 2002
+    +0, +0,  // 2003
+    +0, +0,  // 2004
+    +0, +1,  // 2005
+    +0, +0,  // 2006
+    +0, +0,  // 2007
+    +0, +1,  // 2008
+    +0, +0,  // 2009
+    +0, +0,  // 2010
+    +0, +0,  // 2011
+    +1, +0,  // 2012
+    +0, +0,  // 2013
+    +0, +0,  // 2014
+    +1, +0,  // 2015
+    +0,      // 2016
+}};
+
+// Returns UTC - TAI on the given UTC day (similar to Bulletin C).
+constexpr Time UTC_TAI(Date const& utc_date) {
+  return utc_date.month() == 1 && utc_date.day() == 1
+             ? utc_date.year() == 1972
+                   ? -10 * Second
+                   : -leap_seconds[(utc_date.year() - 1973) * 2] * Second +
+                     -leap_seconds[(utc_date.year() - 1973) * 2 + 1] * Second +
+                     UTC_TAI(Date::Calendar(utc_date.year() - 1, 1, 1))
+             : (utc_date.month() > 6
+                    ? -leap_seconds[(utc_date.year() - 1972) * 2] * Second
+                    : 0 * Second) +
+               UTC_TAI(Date::Calendar(utc_date.year(), 1, 1));
+}
+
+// NOTE(egg): no check for invalid UTC in case of negative leap seconds.
+constexpr Instant DateTimeAsUTC(DateTime const& date_time) {
+  return date_time.time().is_end_of_day()
+             ? DateTimeAsUTC(date_time.normalized_end_of_day())
+             : CHECKING(
+                   !date_time.time().is_leap_second() ||
+                   (date_time.date().month() == 6 &&
+                    leap_seconds[(date_time.date().year() - 1972) * 2] == +1) ||
+                   (date_time.date().month() == 12 &&
+                    leap_seconds[(date_time.date().year() - 1972) * 2 + 1] ==
+                        +1),
+                   DateTimeAsTAIUnchecked(date_time) -
+                       UTC_TAI(date_time.date()));
+}
+
+// |Instant| date literals.
+
+#if (PRINCIPIA_COMPILER_CLANG || PRINCIPIA_COMPILER_CLANG_CL) && WE_LIKE_N3599
+template<typename C, C... string>
+constexpr std::array<C, sizeof...(string)> unpack_as_array() {
+  return std::array<C, sizeof...(string)>{{string...}};
+}
+
+template<typename T>
+constexpr T const& as_const_ref(T const& t) {
+  return t;
+}
+
+template<typename C, std::size_t size>
+constexpr C const* c_str(std::array<C, size> const& array) {
+  // In C++17 this could be |data()|.  For the moment this does the job.
+  return &as_const_ref(array)[0];
+}
+
+template<typename C, C... string>
+constexpr Instant operator""_TAI() {
+  constexpr auto result = DateTimeAsTAI(
```

Comment:
To force constexprness, otherwise if the result is a `CHECK` failure and is not used in a constant expression, the failure will occur at runtime.
Commenting.